### ISSUE TITLE
Improve C preprocessor

### ISSUE THREAD

**Issue by LPeter1997** [LanguageDev/Yoakke#34]
There are a couple of things missing from the preprocessor:
 * [ ] `#if` and `#elif` arithmetic expressions that support `defined()` in them
 * [ ] Directives ignoring line-trailing tokens (make that a setting even?)
 * [ ] Proper ranges for substitution
 * [ ] `#include`s both with string literals and `<>`
 * [ ] Configurable macro argument expansion
   * We could implement `defined` as a custom macro that disables argument expansion
 * [ ] Proper error handling, error events
 * [ ] Warning events (like trailing tokens for directives)
 * [ ] Relay macro definitions that can take delegates
 * Some common pre-defined macros:
   * [ ] `__LINE__`
   * [ ] `__FILE__`
   * [ ] `__COUNTER__`
   * anything else?

**Comment by kant2002** [comment]
Have a question. When C preprocessor works, seems to be information about file parsing context is lost. That's make harder to produce proper diagnostic message if error somewhere inside header file.

At least here in [Cesium](https://github.com/forneVeR/Cesium) we just include strings because Lexer seems to does not know how to navigate multiple files. Am I miss something or maybe there way how this can be done better?

**Comment by LPeter1997** [comment]
@kant2002 Indeed, this is an oversight on my part. The plan was to be able to track where different tokens came from by the [ExpandedFrom](https://github.com/LanguageDev/Yoakke/blob/master/Sources/Core/Yoakke.C.Syntax/CToken.cs#L38) attribute. The problem is that tokens in general don't include which file they came from. I'm planning to change that in the future, to make tokens have a [Location](https://github.com/LanguageDev/Yoakke/blob/master/Sources/Core/Yoakke.Text/Location.cs) instead. I probably didn't notice this, because the [pre-processor I have implemented](https://github.com/LanguageDev/Yoakke/blob/master/Sources/Core/Yoakke.C.Syntax/CPreProcessor.cs) didn't get to includes yet, so it didn't even occur to me that right now there is no decent way to track context.

Thanks for this feedback, I'll try to sort this out as soon as I get the time. Any suggestion around the API is welcome, I'd like to know if you have any "ideal way" of dealing with this.

**Comment by kant2002** [comment]
I do not have answer for that based on current status of parser. I definitely would think about that. My perspective would be based mostly on Roslyn, and maybe on TypeScript, since this is most advanced compiler frameworks in regards to tooling support. Just so you know where how my opinion shaped. I will come back when I have something which sounds like a plan.
